Change in persisting phone number.

binh-dam-ibigroup opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR fixes #79.

Per team discussion, the `/api/secure/user/<id>/verify_sms` GET endpoint (for requesting an SMS) now takes a phone number as path parameter and persists that number as the user's phone number. It also sets `isPhoneNumberVerified = false` ~and `notificationChannel = "sms"` (let me know if we should not set the notification channel)~. A test has been added to check these assignments.

The `/api/secure/user/<id>/verify_sms` POST endpoint now sets `isPhoneNumberVerified = true` ~and `notificationChannel = "sms"`~ upon receiving an "approved" phone verification status.

binh-dam-ibigroup committed:
fix(OtpUserController): Add phone number as path param, save it before sending SMS.

binh-dam-ibigroup committed:
fix(OtpUserController): Set isPhoneNumberVerified after receiving validation check.

binh-dam-ibigroup committed:
refactor(ApiUserFlowRequest): Remove unused var, add phone number comment.

binh-dam-ibigroup committed:
docs(OtpUserController): Add more method comments.

binh-dam-ibigroup committed:
chore(swagger): Update version-controlled swagger.

landonreed commented on src/test/java/org/opentripplanner/middleware/ApiUserFlowTest.java:
```diff
@@ -191,4 +194,48 @@ public void canSimulateApiUserFlow() {
+     * Tests for the chain of API calls to verify a user's phone number.
+     */
+    @Test
+    public void canSimulatePhoneVerificationFlow() {
```
I don't think this test really belongs here. This test class was intended to model how a third party API user interacts with otp-middleware. And an API user (eventually) won't be able to create an OtpUser that can register a phone number for SMS notifications.

binh-dam-ibigroup replied:
I moved and changed the test in https://github.com/ibi-group/otp-middleware/pull/83/commits/44864ac4f129dd0e1bc766444a85c73776583201.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -125,7 +141,17 @@ public VerificationResult verifyPhoneWithCode(Request req, Response res) {
+            // (set isPhoneNumberVerified and notificationChannel).
+            // TODO: Figure out adding a test for this update to the OtpUser record.
+            otpUser.isPhoneNumberVerified = true;
+            otpUser.notificationChannel = "sms";
```
I don't think we should be setting channel here. I think that should be independent of this action (especially considering that we intend to allow a "both" option for notification channel).

binh-dam-ibigroup replied:
Removed. I was on the fence about that.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -88,13 +91,24 @@ protected OtpUser getUserProfile(Auth0UserProfile profile) {
+        // Update OtpUser before submitting SMS request.
+        otpUser.phoneNumber = phoneNumber;
+        otpUser.isPhoneNumberVerified = false;
+        otpUser.notificationChannel = "sms";
```
As mentioned below, I'm not sure if we want to set this automatically (considering the "both" option)?

binh-dam-ibigroup replied:
Removed.

landonreed reviewed (changes requested):
Looks good. Just a couple of comments (location of test and automatically setting notification channel).

binh-dam-ibigroup committed:
refactor(OtpUserControllerTest): Move phone API test to this class. Do not set the notification chan

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -88,23 +91,35 @@ protected OtpUser getUserProfile(Auth0UserProfile profile) {
         // Verification result will show "pending" status if verification text is successfully sent.
         return new VerificationResult(verification);
+ */
+        return null;
```
I'm not sure what's happening here... returning null doesn't seem too useful in this scenario! I'm guessing this is just a temporary comment that accidentally got checked in?

binh-dam-ibigroup replied:
Yes... thanks for catching that.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -88,23 +91,35 @@ protected OtpUser getUserProfile(Auth0UserProfile profile) {
+        // Get phone number from the path param.
+        String phoneNumber = req.params(PHONE_PARAM);
+        if (StringUtils.isBlank(phoneNumber)) {
+            logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "A phone number must be provided.");
```
After further inspection, this should probably validate the phone number against a regex (or some other validator) instead of just checking whether it's blank.

binh-dam-ibigroup replied:
I think Twilio throws an error if the number is not in E.164 (e.g. doesn't have the +1 prefix for US numbers).

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -125,7 +140,16 @@ public VerificationResult verifyPhoneWithCode(Request req, Response res) {
+        VerificationResult verificationResult = new VerificationResult(check);
+        if (verificationResult.status.equals("approved")) {
+            // If the check is successful, update the OtpUser
+            // (set isPhoneNumberVerified and notificationChannel).
```
Probably ok to remove the second line of this comment.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -58,20 +60,21 @@ OtpUser preCreateHook(OtpUser user, Request req) {
                     .withDescription("Request an SMS verification to be sent to an OtpUser's phone number.")
-                    .withPathParam().withName(ID_PARAM).withDescription("The id of the OtpUser.").and()
+                    .withPathParam().withName(ID_PARAM).withRequired(true).withDescription("The id of the OtpUser.").and()
+                    .withPathParam().withName(PHONE_PARAM).withRequired(true).withDescription("The phone number to validate, in E.164 format.").and()
```
While precise, this isn't super helpful for most people. It would be good to add an example to this description (e.g., `+15555551234`).

landonreed replied:
This item is unresolved.

binh-dam-ibigroup replied:
Fixed. in https://github.com/ibi-group/otp-middleware/pull/83/commits/e1e2cd22c54f645d30c6662dbf1e8e366797c6ce

landonreed replied:
Sorry, I didn't mean to remove that it should be E.164 (which is nice because it is precise), just that we should add the example (for folks who don't immediately know what that means). To further clarify: we should be picky about the format used for this phone number field. If it doesn't match e.164, we should reject the request.

It looks like you have added a configurable regex for making this format flexible. I disagree with this change. We should just use a hard-coded regex and enforce a single format in requests.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -88,23 +91,35 @@ protected OtpUser getUserProfile(Auth0UserProfile profile) {
-        Verification verification = NotificationUtils.sendVerificationText(otpUser.phoneNumber);
+
+        // Update OtpUser before submitting SMS request.
+        otpUser.phoneNumber = phoneNumber;
```
This approach overwrites the existing phone number before it is verified. If we do things this way, it will be impossible to send off alerts to the phone until it is verified. And if the user decides they actually don't want to verify after all and keep their old number, we will have overwritten it and they would have to go back in and reconfirm their existing phone number.

The temporary phone number should be stored in a temporary field until it is verified.

binh-dam-ibigroup replied:
See the Revert Number button in https://github.com/opentripplanner/otp-react-redux/pull/224 for part of this concern.

evansiroky replied:
I disagree with the approach of having a revert option due to some of the concerns shared above. Additionally, it is more code to maintain than keeping an existing phone number intact until verification and requiring the user to reverify an original number if they want to change it back.

binh-dam-ibigroup committed:
refactor(OtpUserController): Remove temp comment.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -125,7 +140,16 @@ public VerificationResult verifyPhoneWithCode(Request req, Response res) {
+            // If the check is successful, update the OtpUser
+            // (set isPhoneNumberVerified and notificationChannel).
+            // TODO: Figure out adding a test for this update to the OtpUser record.
+            otpUser.isPhoneNumberVerified = true;
```
Only at this point should the `otpUser.phoneNumber` be updated.

binh-dam-ibigroup replied:
@landonreed and @evansiroky please agree on the above comment. https://github.com/ibi-group/otp-middleware/pull/83/commits/5621a131b26f784d618e0a8cb228ba3e3700c787 does per above, https://github.com/ibi-group/otp-middleware/pull/83/commits/d88766150d215e265639e0525afce51be7c6ec66 does not.

landonreed replied:
Using `isPhoneNumberVerified` is how we originally designed things and in my opinion it works just fine. If there needs to be further discussion or an alternative approach, that should be handled outside of this PR.

evansiroky replied:
I still disagree for reasons mentioned in company chat, but don't care enough to block progress.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -88,23 +91,35 @@ protected OtpUser getUserProfile(Auth0UserProfile profile) {
-        if (otpUser.phoneNumber == null) {
-            logMessageAndHalt(req, HttpStatus.NOT_FOUND_404, "User must have valid phone number to verify.");
+        // Get phone number from the path param.
+        String phoneNumber = req.params(PHONE_PARAM);
```
I'm wondering now for before we send the verification text, if there is a Twilio service we can call to check whether the phone number is already verified by Twilio?

binh-dam-ibigroup replied:
Twilio offers [caller ID management for outgoing calls](https://support.twilio.com/hc/en-us/articles/223180048-Adding-a-Verified-Phone-Number-or-Caller-ID-with-Twilio), but that is not quite what you are talking about.

binh-dam-ibigroup replied:
Actually it is, "outgoing" meaning the destination number a message goes out to... Do you want to address it in this PR or another one?

landonreed replied:
I was just curious what was out there, no need to address.

landonreed reviewed (changes requested):
Just a few more items I ran across while testing otp-react-redux.

evansiroky reviewed (changes requested):
Please refactor so that an existing confirmed phone number does not get overwritten. See https://github.com/ibi-group/otp-middleware/pull/83#discussion_r502023720

binh-dam-ibigroup committed:
fix(OtpUserController): Use new OtpUser.pendingPhoneNumber to hold unverified number.

BREAKING CHANGE: OtpUser.isPhoneNumberVerified is removed. OtpUser.pendingPhoneNumber is for
unverified numbers and OtpUser.phoneNumber should only hold verified numbers.

binh-dam-ibigroup committed:
fix(OtpUser): Add phone format verification and pending numbers.

binh-dam-ibigroup committed:
test(OtpUserControllerTest): Add test to flag invalid/foreign numbers.

binh-dam-ibigroup committed:
chore(swagger): Update version-controlled swagger (for CI test).

landonreed commented on src/main/java/org/opentripplanner/middleware/utils/NotificationUtils.java:
```diff
@@ -157,5 +165,77 @@ public static boolean sendEmailViaSendGrid(String toEmail, String subject, Strin
+     * Ensures that the provided phone number is a domestic mobile number.
+     * Returns a 400-Bad request status if that is not the case.
+     */
+    public static com.twilio.rest.lookups.v1.PhoneNumber ensureDomesticPhoneNumber(spark.Request req, String phoneNumberString) {
```
This method needs a lot of work before being ready to ship:

- First, I don't think it should take a request as a parameter. Take a look at `ApiGatewayUtils#createApiKey` which throws a `CreateApiKeyException` for a pattern that could be mimicked here.

https://github.com/ibi-group/otp-middleware/blob/692c66afa6e2af09e415154039d870a556ff2260/src/main/java/org/opentripplanner/middleware/utils/ApiGatewayUtils.java#L64

- Also, it appears to contain some strange logic for fake numbers, which could cause issues in a production environment.
- It also uses a Twilio service (lookup API) to perform this lookup. I think a simpler (and less costly) approach would just to use a regex as I described in a previous comment.

In summary, I don't think this method needs to exist at all. Please just implement a regex check as mentioned previously: https://stackoverflow.com/a/23299989/915811

binh-dam-ibigroup replied:
I see three big advantages for using Twilio's lookup API over a regex, if cost is not too much of a concern:
- No need to configure a regex for each phone number format we encounter (the country code is enough)
- We can accept phone numbers from the UI in any format, and Twilio will convert it into both an E.164 format (+15555550123) to use to send trip monitoring notifications, and a localized format (555) 555-0123 that is sent back to the UI as of writing. That also means we don't have to implement the logic/configurations to strip the extra characters or figure out country prefixes in order to create an E.164 number for notifications.
- We can exclude international numbers from notifications (for US-formatted numbers, that including Canadian and non-US numbers that have the +1 prefix), and potentially landlines.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -88,13 +92,25 @@ protected OtpUser getUserProfile(Auth0UserProfile profile) {
+        // Update OtpUser.pendingPhoneNumber before submitting SMS request.
+        String phoneNumberToSubmit = validNumber.getPhoneNumber().toString();
+        otpUser.pendingPhoneNumber = phoneNumberToSubmit;
+        otpUser.pendingPhoneNumberFormatted = validNumber.getNationalFormat();
```
This is redundant information.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -113,19 +130,29 @@ public VerificationResult verifyPhoneWithCode(Request req, Response res) {
+            // If the check is successful, update the OtpUser's phoneNumber and erase pendingPhoneNumber.
+            // TODO: Figure out adding a test for this update to the OtpUser record.
+            otpUser.phoneNumber = otpUser.pendingPhoneNumber;
+            otpUser.pendingPhoneNumber = null;
```
We did not agree to implement this change from storing a single phone number to pending + phone number.

binh-dam-ibigroup replied:
This is to not overwrite an existing confirmed number.

landonreed reviewed (changes requested):
We did not agree to make the change to OtpUser to add a `pendingPhoneNumber` field. This also adds what appears to be an overly-complex validation method to NotificationUtils that should likely be removed.

binh-dam-ibigroup committed:
refactor(NotificationUtils): Erase pendingPhoneNumber before checking new phone validity.

binh-dam-ibigroup committed:
refactor(OtpUserController): Remove pendingPhonenumber, use regex to check numbers.

binh-dam-ibigroup committed:
docs(OtpUserController): Update resource swagger description

landonreed commented on configurations/default/env.yml.tmp:
```diff
@@ -17,6 +17,11 @@ AWS_API_SERVER: aws-api-id.execute-api.us-east-1.amazonaws.com
+### Optional phone number prefix (defaults to "+1" [North America])
+# PHONE_COUNTRY_PREFIX: "+1"
+### Optional phone number regexp to enable phone number checking.
+# PHONE_NUMBER_REGEXP: ^\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$
```
This should be hardcoded and check for E.164 format

binh-dam-ibigroup replied:
Removed in https://github.com/ibi-group/otp-middleware/pull/83/commits/a9386a6e2f46bc9529e14d51ff9e51106ec87ba6

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -98,16 +98,19 @@ public VerificationResult sendVerificationText(Request req, Response res) {
-        otpUser.pendingPhoneNumber = phoneNumberToSubmit;
-        otpUser.pendingPhoneNumberFormatted = validNumber.getNationalFormat();
+        // Update OtpUser.phoneNumber before submitting SMS request.
+        String phoneNumberToSubmit = NotificationUtils.getRawPhoneNumber(phoneNumber);
```
We should expect clients/requests to contain the country code. Your inclination to make this E.164 format was correct, I'm not sure why you deviated from this. I don't think we really need the config value for country code in this case.

binh-dam-ibigroup replied:
I can make the UI add the country prefix if that simplifies the backend. It is odd to ask regular users to enter a +1 every time.

landonreed replied:
Yep, I think it's more appropriate for the UI to supply this.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/OtpUser.java:
```diff
@@ -28,23 +31,10 @@
-
-    /**
-     * Verified phone number for SMS notifications, in +15551234 format (E.164 format, includes country code, no spaces).
+     * Phone number for SMS notifications.
```
Old comment about E.164 format is correct and should be reverted. We want a predictable consistent way for clients to interact with this API. And enforcing that format is an important part of that in my view.

binh-dam-ibigroup replied:
Reverted in https://github.com/ibi-group/otp-middleware/pull/83/commits/74075ba31332186ab2906e45a2b02c3ff7906448

landonreed commented on src/main/java/org/opentripplanner/middleware/models/OtpUser.java:
```diff
@@ -28,23 +31,10 @@
-    /**
-     * Locally-formatted phoneNumber for display.
-     */
-    public String phoneNumberFormatted;
```
Thanks for removing. If we want some special formatting, we should just handle on the client side with an npm lib.

landonreed commented on src/main/java/org/opentripplanner/middleware/utils/NotificationUtils.java:
```diff
@@ -167,75 +168,26 @@ public static boolean sendEmailViaSendGrid(String toEmail, String subject, Strin
-            );
-        }
+        // If no regex string/pattern is configured, assume phone number is valid.
+        if (PHONE_PATTERN == null) return true;
```
This will go away once we hardcode the regex.

binh-dam-ibigroup replied:
I don't think we want to hardcode the regex.

landonreed commented on src/main/java/org/opentripplanner/middleware/utils/NotificationUtils.java:
```diff
@@ -167,75 +168,26 @@ public static boolean sendEmailViaSendGrid(String toEmail, String subject, Strin
+     * by stripping non-digit characters and prefixing with the configured country code.
+     */
+    public static String getRawPhoneNumber(String formattedNumber) {
+        return PHONE_COUNTRY_PREFIX + formattedNumber.replaceAll("\\D", "");
```
I think this can go away once we enforce that clients pass phone numbers in E.164.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/83/commits/a9386a6e2f46bc9529e14d51ff9e51106ec87ba6.

landonreed reviewed (changes requested):
Thanks for the changes, @binh-dam-ibigroup. I think the only thing left is to enforce E.164 format in requests. I'm not sure if something I said caused you to deviate from how you had originally designed things. If so, I apologize -- I think your instincts were correct!

evansiroky commented on src/test/java/org/opentripplanner/middleware/controllers/api/OtpUserControllerTest.java:
```diff
@@ -0,0 +1,128 @@
+        assertTrue(otpUserWithPhone.isPhoneNumberVerified);
+    }
+
+    @ParameterizedTest
+    @MethodSource("createPhoneNumberTestCases")
+    public void isPhoneNumberValid(Map.Entry<String, Boolean> testCase) {
+        assumeTrue(testsShouldRun());
+        String number = testCase.getKey();
+        boolean isValid = testCase.getValue();
+
+        assertEquals(isValid, NotificationUtils.isPhoneNumberIsValid(number));
+    }
+
+    private static Set<Map.Entry<String, Boolean>> createPhoneNumberTestCases() {
+        HashMap<String, Boolean> cases = new HashMap<>();
+        cases.put("5555550123", true);
+        cases.put("(555) 555-0123", true);
+        cases.put("555 555 0123", true);
+        cases.put("(555) 555,0123", false);
+        cases.put("555555", false);
+        cases.put("55555555555555", false);
+
+        return cases.entrySet();
+    }
+
+    @Test
+    public void getE164Number() {
+        assumeTrue(testsShouldRun());
+        String formattedNumber = "(800) 555-0123";
+        assertEquals("+18005550123", NotificationUtils.getRawPhoneNumber(formattedNumber));
+    }
```
These 2 tests are for methods entirely contained within the NotificationUtils class, so they should be placed in a the NotificationUtilsTest class or it should be created if it doesn't already exist.

binh-dam-ibigroup replied:
I moved the check to `OtpUserController` in https://github.com/ibi-group/otp-middleware/pull/83/commits/a9386a6e2f46bc9529e14d51ff9e51106ec87ba6

evansiroky reviewed (approved):
Contingent on https://github.com/ibi-group/otp-middleware/pull/83#discussion_r505918524 and @landonreed's approval following E.164 regex hardcoding check.

Still disagree about overwritten a confirmed number with an unconfirmed one, but don't care enough to block progress.

binh-dam-ibigroup committed:
refactor(OtpUserController): Accept E.164 phone numbers only.

binh-dam-ibigroup committed:
docs(OtpUser/Controller): Tweak comments

binh-dam-ibigroup committed:
fix(OtpUserController): Persist pending number after SMS sent.

binh-dam-ibigroup committed:
refactor(OtpUserControllerTest): Remove unused imports

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -130,11 +140,11 @@ public VerificationResult verifyPhoneWithCode(Request req, Response res) {
         }
-        if (otpUser.pendingPhoneNumber == null) {
-            logMessageAndHalt(req, 404, "User must have valid phone number for SMS verification.");
+        if (StringUtils.isBlank(otpUser.phoneNumber)) {
```
Is there a good reason to use `isBlank` instead of `isPhoneNumberValidE164`?

binh-dam-ibigroup replied:
It's evolution!

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/83/commits/8ec525aabdaac519033d9918bca89ad483ecaa7f.

landonreed commented on src/test/java/org/opentripplanner/middleware/controllers/api/OtpUserControllerTest.java:
```diff
@@ -60,69 +61,70 @@ public static void tearDown() {
+    @ParameterizedTest
+    @MethodSource("createBadPhoneNumbers")
+    public void invalidNumbersShouldProduceBadRequest(Map.Entry<String, Integer> testCase) {
+        assumeTrue(testsShouldRun());
```
I feel like there is a lot of churn happening on this auth disabled stuff, but I'm pretty sure in some PR we started to use the pattern of not checking `DISABLE_AUTH=false` but rather setting it with Auth0Connection#setAuthDisabled and reverting the value to getDefaultAuthDisabled in tearDown. Please follow that pattern here. Also, does e2e actually need to be enabled for these tests to work?

binh-dam-ibigroup replied:
Good point. e2e is not required.

binh-dam-ibigroup replied:
Updated in https://github.com/ibi-group/otp-middleware/pull/83/commits/8ec525aabdaac519033d9918bca89ad483ecaa7f.

landonreed commented on src/test/java/org/opentripplanner/middleware/controllers/api/OtpUserControllerTest.java:
```diff
@@ -60,69 +61,70 @@ public static void tearDown() {
+    }
+
+    private static Set<Map.Entry<String, Integer>> createBadPhoneNumbers() {
+        HashMap<String, Integer> cases = new HashMap<>();
```
I kind of like the parameterized pattern that uses `Arguments`. That approach allows for the test args to be `String phoneNumber, int statusCode`.

```java
private static Stream<Arguments> createBadPhoneNumbers() {
        return Stream.of(
            Arguments.of("5555555", HttpStatus.BAD_REQUEST_400),
            Arguments.of("+15555550001", HttpStatus.INTERNAL_SERVER_ERROR_500)
        );
    }
```

landonreed commented on src/test/java/org/opentripplanner/middleware/controllers/api/OtpUserControllerTest.java:
```diff
@@ -60,69 +61,70 @@ public static void tearDown() {
+        cases.put("+15555550123", true);
+        cases.put("+1 5555550123", false); // no spaces allowed.
+        cases.put("(555) 555,0123", false);
+        cases.put("555555", false);
```
Same comment about Arguments.

binh-dam-ibigroup replied:
Good find, it's a cleaner syntax. Updated in https://github.com/ibi-group/otp-middleware/pull/83/commits/8ec525aabdaac519033d9918bca89ad483ecaa7f.

landonreed reviewed (approved):
Just a couple of comments that could be addressed, but otherwise looks great.

binh-dam-ibigroup committed:
refactor(OtpUserController/Test): Address PR comments

binh-dam-ibigroup committed:
refactor(OtpUserController): Tweak message when submitting code without valid phone number.

binh-dam-ibigroup committed:
fix(OtpUserController): Add missing ! in number check when sending code.

binh-dam-ibigroup committed:
refactor(OtpUserController): Make SMS the notification channel when requesting SMS.
